core/Enclave/key_operation.h

```diff
@@ -0,0 +1,123 @@
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+#include "log_utils.h"
+#include "sgx_tseal.h"
+
+#include <string>
+#include <stdio.h>
+#include <stdbool.h>
+#include <mbusafecrt.h>
+
+#include "sgx_report.h"
+#include "sgx_utils.h"
+#include "sgx_tkey_exchange.h"
+
+#ifndef _KEY_OPERATION_H_
+#define _KEY_OPERATION_H_
+
+void printf(const char *fmt, ...);
+
+sgx_status_t ehsm_aes_gcm_encrypt(const uint8_t *aad, size_t aad_len,
+                                  const uint8_t *cmk_blob, size_t SIZE_OF_KEYBLOB_T,
+                                  const uint8_t *plaintext, size_t plaintext_len,
+                                  uint8_t *cipherblob, size_t cipherblob_len,
+                                  ehsm_keyspec_t keyspec);
+
+sgx_status_t ehsm_aes_gcm_decrypt(const uint8_t *aad, size_t aad_len,
+                                  const uint8_t *cmk_blob, size_t SIZE_OF_KEYBLOB_T,
+                                  const uint8_t *cipherblob, size_t cipherblob_len,
+                                  uint8_t *plaintext, size_t plaintext_len,
+                                  ehsm_keyspec_t keyspec);
+
+sgx_status_t ehsm_sm4_encrypt(const uint8_t *aad, size_t aad_len,
+                              const uint8_t *cmk_blob, size_t SIZE_OF_KEYBLOB_T,
+                              const uint8_t *plaintext, size_t plaintext_len,
+                              uint8_t *cipherblob, size_t cipherblob_len,
+                              ehsm_keyspec_t keyspec);
+
+sgx_status_t ehsm_sm4_decrypt(const uint8_t *aad, size_t aad_len,
+                              const uint8_t *cmk_blob, size_t SIZE_OF_KEYBLOB_T,
+                              const uint8_t *cipherblob, size_t cipherblob_len,
+                              uint8_t *plaintext, size_t plaintext_len,
+                              ehsm_keyspec_t keyspec);
+
+sgx_status_t ehsm_asymmetric_encrypt(const ehsm_keyblob_t *cmk, ehsm_data_t *plaintext, ehsm_data_t *ciphertext);
+
+sgx_status_t ehsm_asymmetric_decrypt(const ehsm_keyblob_t *cmk, ehsm_data_t *ciphertext, ehsm_data_t *plaintext);
+
+sgx_status_t ehsm_rsa_sign(const ehsm_keyblob_t *cmk_blob,
+                           ehsm_padding_mode_t padding_mode,
+                           ehsm_digest_mode_t digest_mode,
+                           ehsm_keyspec_t keyspec,
```

Comment:
why add the keyspec, it should already included in the cmk_blob

Reply:
not see the changes, pading, digest_mode, keyspec should already in the cmk struct

Reply:
it will be changed